File: cpp/heapSort.cpp
#include <iostream>

using namespace std;

void heapify(int arr[], int n, int i) {
  int largest = i;
  int l = 2 * i + 1;
  int r = 2 * i + 2;

  // if left child > root
  if(l < n && arr[l] > arr[largest])
    largest = l;

  // if right child > largest so far
  if(r < n && arr[r] > arr[largest])
    largest = r;

  // if largest is not root
  if(largest != i) {
    swap(arr[i], arr[largest]);
    heapify(arr, n, largest);
  }
}

// main function to do heapsort
void heapSort(int arr[], int n) {
  // build heap (rearrange array)
  for(int i = n / 2 - 1; i >= 0; i--)
    heapify(arr, n, i);

  // one by one extract element from heap
  for(int i = n - 1; i > 0; i--) {
    swap(arr[0], arr[i]);
    heapify(arr, i, 0);
  }
}
// function to print array
void printArray(int arr[], int n) {
  for(int i = 0; i < n; i++)
    cout << arr[i] << " ";
  cout << "\n";
}

// driver program
int main() {
  int arr[] = {12, 11, 13, 5, 6, 7};
  int n = sizeof(arr) / sizeof(arr[0]);
  heapSort(arr, n);
  cout << "sorted array is \n";
  printArray(arr, n);
  return 0;
}

File: python/heapSort.py
def heapify(arr, n, i):
  largest = i         # initialize largest as root
  l = 2 * i + 1  
  r = 2 * i + 2

  # see if left child of root exists and is > root
  if l < n and arr[l] > arr[largest]:
    largest = 1

  # see if right child of root exists and is > largest so far
  if r < n and arr[r] > arr[largest]:
    largest = r

  # change root if needed
  if largest != i:
    arr[i], arr[largest] = arr[largest], arr[i]     # swap
    # heapify the root
    heapify(arr, n, largest)


def heapSort(arr):
  n = len(arr)
  # build a maxheap
  for i in range(n // 2 - 1, -1, -1):
    heapify(arr, n, i)

  #one by one extract element from heap
  for i in range(n - 1, 0, -1):
    # move current root to end
    arr[i], arr[0] = arr[0], arr[i]

    # call max heapify on reduced heap
    heapify(arr, i, 0)

# example usage
arr = [12, 11, 13, 5, 6, 7]
heapSort(arr)
n = len(arr)
print("sorted array is")
for i in range(n):
  print("%d " %arr[i], end='')
  
